### PR TITLE
[JENKINS-21017] When unmarshalling into an existing object, reset missing fields

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-21017](https://issues.jenkins-ci.org/browse/JENKINS-21017)

@jenkinsci/code-reviewers @reviewbybees

Notes:

* Likely we will need a good bit of testing before we consider merging
* Needs at least a full clean test run and a full ATH run before I am ready to claim this is even close to mergable

### CODE REVIEW

File: core/src/main/java/hudson/util/RobustReflectionConverter.java [hunk 1]
```diff
@@ -371,6 +442,12 @@ public Object doUnmarshal(final Object result, final HierarchicalStreamReader re
             OldDataMonitor.report((Saveable)result, (ArrayList<Throwable>)context.get("ReadError"));
             context.put("ReadError", null);
         }
+        if (existingObject) {
+            for (Map.Entry<FieldExpectation, Object> entry : expectedFields.entrySet()) {
+                reflectionProvider.writeField(result, entry.getKey().getName(), entry.getValue(),
```
Comment (jglick) [inline core/src/main/java/hudson/util/RobustReflectionConverter.java]: Does this wind up writing non-null fields of root objects twice?

**Review by jglick — APPROVED** [review]
Seems right. I think the key advance here was in finding an existing general-purpose `Converter` defined in Jenkins sources whose behavior could be patched easily.

Would be advisable to also write an integration test for one of the originally reported systems, such as trying to unset an `AbstractProject.assignedNode` by `POST`ing to its `config.xml`.

File: core/src/test/java/hudson/util/XStream2Test.java [hunk 2]
```diff
@@ -296,4 +300,218 @@ public void trimVersion() {
...
+
+        String xmlA = Jenkins.XSTREAM2.toXML(Jenkins.XSTREAM2.fromXML(emptyXml, existingInstance));
+        String xmlB = Jenkins.XSTREAM2.toXML(Jenkins.XSTREAM2.fromXML(emptyXml, newInstance));
+        String xmlC = Jenkins.XSTREAM2.toXML(Jenkins.XSTREAM2.fromXML(emptyXml, null));
+
+        assertThat("Deserializing over an existing instance is the same as with no root", xmlA, is(xmlC));
+        assertThat("Deserializing over an new instance is the same as with no root", xmlB, is(xmlC));
+    }
+
+    public static class WithDefaults {
```
Comment (oleg-nenashev) [inline core/src/test/java/hudson/util/XStream2Test.java]: Builder pttern would be preferable

**Review by oleg-nenashev — APPROVED** [review]
LGTM though I have not tested it

File: core/src/main/java/hudson/util/RobustReflectionConverter.java (hunk 1 above)
Comment (stephenc) [inline core/src/main/java/hudson/util/RobustReflectionConverter.java]: No, these are the fields that were not overwritten, and this is only for an already existing object.

I think the issue also affects non root objects only people have got away without noticing because of the root object blocker (or else the child objects are not updated)... need tests to verify further

File: core/src/main/java/hudson/util/RobustReflectionConverter.java [hunk 3]
```diff
@@ -271,8 +271,71 @@ public Object unmarshal(final HierarchicalStreamReader reader, final Unmarshalli
...
+        }
+
+
+    }
+
     public Object doUnmarshal(final Object result, final HierarchicalStreamReader reader, final UnmarshallingContext context) {
         final SeenFields seenFields = new SeenFields();
+        final boolean existingObject = context.currentObject() != null;
+        final Map<FieldExpectation, Object> expectedFields = existingObject ? new HashMap<FieldExpectation, Object>() : null;
+        final Object cleanInstance = existingObject ? reflectionProvider.newInstance(result.getClass()) : null;
```
Comment (daniel-beck) [inline core/src/main/java/hudson/util/RobustReflectionConverter.java]: This and the preceding line would be clearer if they just initialized to `null` and were populated in the `if` block.

**Review by daniel-beck — APPROVED** [review]

**Comment by daniel-beck** [comment]
A test using `POST config.xml` and a few top-level elements e.g. in `FreestyleJob` would be useful, e.g. confirming that description would be unset, or that all build steps are removed as expected.

**Comment by jglick** [comment]
> I think the issue also affects non root objects

I doubt it. Everything not at top level is just constructed from scratch. There is no XStream equivalent of `ReconfigurableDescribable` that I know of.

**Comment by stephenc** [comment]
@daspilker yes *but* I now have a lot of NPEs in the test cases deserialising other objects, so I need to investigate what exactly is going on there

File: core/src/main/java/hudson/util/RobustReflectionConverter.java (hunk 3 above)
Comment (daspilker) [inline core/src/main/java/hudson/util/RobustReflectionConverter.java]: `reflectionProvider` always seems to be an instance of `SunLimitedUnsafeReflectionProvider` which uses `sun.misc.Unsafe#allocateInstance` to create a new instance. AFAIK `Unsafe#allocateInstance` does not call any constructor, so fields will not be initialized. So the value of each field in `cleanInstance` will be `null`, `0`, `false`, etc. That's why there are so many NPEs in the tests.

**Comment by lcary** [comment]
@stephenc Any update on this change and/or the NPEs in the test suite? If it fixes [JENKINS-21017](https://issues.jenkins-ci.org/browse/JENKINS-21017), this change should solve many user-reported issues for a jenkins system I'm working on.

**Comment by abayer** [comment]
@stephenc Could you merge from master to this to resolve the conflict?

File: core/src/main/java/hudson/util/RobustReflectionConverter.java (hunk 3 above)
Comment (abayer) [inline core/src/main/java/hudson/util/RobustReflectionConverter.java]: Yeah, a pile of NPEs are specifically coming from https://github.com/stephenc/jenkins/blob/c62f4f753d16210b67ec381b9e5d1d60145594d8/core/src/main/java/jenkins/model/Jenkins.java#L3049, and then if that's bypassed, https://github.com/stephenc/jenkins/blob/c62f4f753d16210b67ec381b9e5d1d60145594d8/core/src/main/java/jenkins/model/Jenkins.java#L3095, and so on. The core problem is that the fields are just not getting initialized, and so boom.

Comment (abayer) [inline core/src/main/java/hudson/util/RobustReflectionConverter.java]: Clarification - the init NPEs go away when I force it to bypass anything involving `clouds` or `views`, which conveniently appear to be the only `final` non-`transient` fields initialized in `Jenkins`. Coincidence? Mayhaps, I dunno.

File: core/src/main/java/hudson/util/RobustReflectionConverter.java (hunk 1 above)
Comment (abayer) [inline core/src/main/java/hudson/util/RobustReflectionConverter.java]: Actually, it looks to me like it does - the expected field value is always uninitialized (see above discussion), and it's not checking whether the field was already seen, so it's overwriting with null/0/false/etc.

**Comment by abayer** [comment]
So I'm trying to figure out the test failures, and the most definite problem is that at least two fields in `Jenkins` (`clouds` and `views`) are ending up with null values rather than what they're initialized with. These two fields are both `final` and not `transient` and seem to be the only fields with that combination of modifiers. I'm also concerned that the `expectedFields` fields are getting overwritten even if they're already set, but not sure on that yet.

**Comment by abayer** [comment]
Verified on the overwriting - the `expectedFields.remove(...)` calls aren't working, because the `classDefiningField` seems to always be `null` (at least with, e.g., the `ClassicPluginStrategyTest` tests) and `result.getClass()` is `hudson.model.Hudson` but the `FieldExpectation.definingClass` is being set to, say, `jenkins.model.Jenkins`, `hudson.model.Node`, `hudson.model.AbstractCIBase`, etc.  So fields like `jdk` are getting written once with the right value and then overwritten with `null`, etc. 

So basically, this is all busted in new and interesting ways.

**Review by abayer — CHANGES_REQUESTED** [review]
👎 until this actually works as intended. Which it does not.

**Comment by abayer** [comment]
@stephenc I do have an experiment at https://github.com/abayer/jenkins/commit/e27e0f252ca180d5abf405b43f7dcb1b6856afc1 that seems to get past at least some of the problems, but I'm not sure if it's just terrible on the face of it. So I defer to you.

**Comment by jglick** [comment]
It would seem to suffice for the `unmarshal(HierarchicalStreamReader, Object)` method to just look for all declared fields in the root object and null them out before calling the regular implementation. Or am I missing something?
